### ISSUE TITLE
Add buttons for backup/restore user settings

### ISSUE THREAD

**Issue by dhowe** [dhowe/AdNauseam#813]
Request from user: https://github.com/dhowe/AdNauseam/issues/810

We currently don't have these buttons. @cqx931 suggests:

> My opinion is that the current Export/Import/Clear-Ads buttons in the settings page are actually duplicates. I’m not sure about other users, but I mostly use these in the vault instead of the settings page. And if having that backup & restore function can help users to backup their settings and rules, I think it is worth the work...

![screen shot 2017-01-24 at 11 15 54 pm](https://cloud.githubusercontent.com/assets/737638/22253128/67149caa-e28b-11e6-9538-3845669e646d.png)

**Comment by dhowe** [comment]
I agree that it is a useful feature, however it should NOT replace our current import/export ads buttons;
perhaps just place below in the interface...

**Comment by cqx931** [comment]
How about something like this?
![image](https://cloud.githubusercontent.com/assets/3056154/24992107/1df3249c-2052-11e7-8385-ebca17bbb687.png)

**Comment by dhowe** [comment]
Or something like this ?

![image](https://cloud.githubusercontent.com/assets/737638/24995249/c3dfb1c8-2060-11e7-94a0-a5fc4aee682a.png)

**Comment by cqx931** [comment]
I just looked into this more in detail and found that it's better to use uBlock's wording for the buttons if we do want to include them in adnauseam. (And I still think it's better to remove the ads buttons from the settings page...)

> Back up to file
> Restore from file... 
> Reset to default settings...

While personal rules/filters could be exported separately and imported, these import buttons in uBlock origin are not only for user settings, but a bundle backup for all the rules and the settings.
Below is an example about how the backup file looks like for uBlock Origin:

```
{
  "timeStamp": 1492163951754,
  "version": "1.11.4",
  "userSettings": {
    "advancedUserEnabled": false,
    "alwaysDetachLogger": false,
    "autoUpdate": true,
    "cloudStorageEnabled": false,
    "collapseBlocked": true,
    "colorBlindFriendly": false,
    "contextMenuEnabled": true,
    "dynamicFilteringEnabled": false,
    "externalLists": "",
    "firewallPaneMinimized": true,
    "hyperlinkAuditingDisabled": true,
    "ignoreGenericCosmeticFilters": false,
    "largeMediaSize": 50,
    "parseAllABPHideFilters": true,
    "prefetchingDisabled": true,
    "requestLogMaxEntries": 1000,
    "showIconBadge": true,
    "tooltipsDisabled": false,
    "webrtcIPAddressHidden": false
  },
  "selectedFilterLists": [
    "plowe-0",
    "malware-1",
    "malware-0",
    "easyprivacy",
    "easylist",
    "ublock-unbreak",
    "ublock-privacy",
    "ublock-badware",
    "ublock-filters",
    "user-filters"
  ],
  "hiddenSettingsString": "assetFetchTimeout 30\nautoUpdateAssetFetchPeriod 120\nautoUpdatePeriod 7\nignoreRedirectFilters false\nignoreScriptInjectFilters false\nmanualUpdateAssetFetchPeriod 2000\npopupFontSize unset\nsuspendTabsUntilReady false",
  "netWhitelist": "about-scheme\nbehind-the-scene\nchrome-extension-scheme\nloopconversation.about-scheme\nmoz-extension-scheme\nopera-scheme\nvivaldi-scheme\nwww.google.com.hk",
  "dynamicFilteringString": "behind-the-scene * 3p-frame noop\nbehind-the-scene * 3p noop",
  "urlFilteringString": "",
  "hostnameSwitchesString": "",
  "userFilters": "",
  "filterLists": {
    "assets/thirdparties/pgl.yoyo.org/as/serverlist": {
      "off": false
    },
    "assets/thirdparties/mirror1.malwaredomains.com/files/justdomains": {
      "off": false
    },
    "assets/thirdparties/www.malwaredomainlist.com/hostslist/hosts.txt": {
      "off": false
    },
    "assets/thirdparties/easylist-downloads.adblockplus.org/easyprivacy.txt": {
      "off": false
    },
    "assets/thirdparties/easylist-downloads.adblockplus.org/easylist.txt": {
      "off": false
    },
    "assets/ublock/unbreak.txt": {
      "off": false
    },
    "assets/ublock/privacy.txt": {
      "off": false
    },
    "assets/ublock/badware.txt": {
      "off": false
    },
    "assets/ublock/filters.txt": {
      "off": false
    },
    "assets/user/filters.txt": {
      "off": false
    },
    "assets/ublock/experimental.txt": {
      "off": true
    }
  }
}
```

**Comment by dhowe** [comment]
makes sense, however we will have to update the function that creates this data to include our user-settings as well

**Comment by cqx931** [comment]
I can first make sure that the function creates the correct data and we can add the buttons back afterwards.

**Comment by cqx931** [comment]
Right now it prints out both the user settings as well as the whole admap to the file.
If we are going to seperate these two export, admap data should be removed from the exported file.
Other than that the data looks ok, it contains all user-settings that are unique to adnauseam.
```
{
  "timeStamp": 1496732838065,
  "version": "3.3.341",
  "userSettings": {
 "parseTextAds": true,
    "eventLogging": true,
    "hidingAds": true,
    "clickingAds": true,
    "blockingMalware": true,
    "disableHidingForDNT": true,
    "disableClickingForDNT": true,
    "clickProbability": 0.75,
    "clickOnlyWhenIdleFor": 0,
    "noIncomingCookies": true,
    "noOutgoingCookies": false,
    "noOutgoingReferer": false,
    "noOutgoingUserAgent": false,
    "advancedUserEnabled": false,
    "alwaysDetachLogger": false,
    "autoUpdate": true,
    "cloudStorageEnabled": false,
    "collapseBlocked": true,
    "colorBlindFriendly": false,
    "contextMenuEnabled": true,
    "dynamicFilteringEnabled": false,
    "externalLists": [],
    "firewallPaneMinimized": true,
    "hyperlinkAuditingDisabled": true,
    "ignoreGenericCosmeticFilters": false,
    "largeMediaSize": 50,
    "parseAllABPHideFilters": true,
    "prefetchingDisabled": true,
    "requestLogMaxEntries": 1000,
    "showIconBadge": true,
    "tooltipsDisabled": false,
    "webrtcIPAddressHidden": false
  },
  "selectedFilterLists": [
    "adnauseam-filters",
    "assets.json",
    "easylist",
    "easyprivacy",
    "eff-dnt-whitelist",
    "malware-0",
    "malware-1",
    "public_suffix_list.dat",
    "ublock-badware",
    "ublock-filters",
    "ublock-privacy",
    "ublock-resources",
    "ublock-unbreak",
    "user-filters"
  ],
  "hiddenSettingsString": "assetFetchTimeout 30\nautoUpdateAssetFetchPeriod 120\nautoUpdatePeriod 7\nignoreRedirectFilters false\nignoreScriptInjectFilters false\nmanualUpdateAssetFetchPeriod 2000\npopupFontSize unset\nsuspendTabsUntilReady false",
  "netWhitelist": "about-scheme\nbehind-the-scene\nchrome-extension-scheme\nchrome-scheme\nloopconversation.about-scheme\nmoz-extension-scheme\nnative.sharethrough.com\nopera-scheme\nsecurepubads.g.doubleclick.net\nvivaldi-scheme",
  "dynamicFilteringString": "behind-the-scene * 3p-frame noop\nbehind-the-scene * 3p noop",
  "urlFilteringString": "",
  "hostnameSwitchesString": "",
  "userFilters": "hongkong.asiaxpat.com###fade\nhongkong.asiaxpat.com###light\n",
  "filterLists": {
    "assets/ublock/adnauseam.txt": {
      "off": false
    },
    "assets.json": {
      "off": false
    },
    "assets/thirdparties/easylist-downloads.adblockplus.org/easylist.txt": {
      "off": false
    },
    "assets/thirdparties/easylist-downloads.adblockplus.org/easyprivacy.txt": {
      "off": false
    },
    "eff-dnt-whitelist": {
      "off": false
    },
    "assets/thirdparties/www.malwaredomainlist.com/hostslist/hosts.txt": {
      "off": false
    },
    "assets/thirdparties/mirror1.malwaredomains.com/files/justdomains": {
      "off": false
    },
    "assets/thirdparties/publicsuffix.org/list/effective_tld_names.dat": {
      "off": false
    },
    "assets/ublock/badware.txt": {
      "off": false
    },
    "assets/ublock/filters.txt": {
      "off": false
    },
    "assets/ublock/privacy.txt": {
      "off": false
    },
    "assets/ublock/resources.txt": {
      "off": false
    },
    "assets/ublock/unbreak.txt": {
      "off": false
    },
    "assets/user/filters.txt": {
      "off": false
    },
    "assets/thirdparties/www.eff.org/files/effdntlist.txt": {
      "off": true
    },
    "assets/ublock/experimental.txt": {
      "off": true
    },
    "assets/thirdparties/pgl.yoyo.org/as/serverlist": {
      "off": true
    }
  }
}
```

**Comment by dhowe** [comment]
Lets put this on hold

**Comment by dhowe** [comment]
lets try to get this into 3.5

**Comment by MrDemocracy** [comment]
I would also like to have this feature. I find it a bit annoying that I have to click 4 export/import buttons, in order to backup/restore all my AdNauseam data. 

Would something like this look good?
<img width="1280" src="https://user-images.githubusercontent.com/23060864/31382942-94236b96-adb9-11e7-805b-5464eb88dd3c.png">

**Comment by dhowe** [comment]
are you suggesting that export-ads only does ads, while backup-to-file does settings and ads?

**Comment by cqx931** [comment]
I think what @MrDemocracy suggests make sense. In this case, the user only need to click two buttons to backup everything.

**Comment by dhowe** [comment]
Seems more convenient, but a bit less clear... Though as you suggest, we could remove ad import/export from this page altogether. Another issue is file size. An ad file can get quite large if there are 10,000+ ads, while a settings backup remains quite small. 

@mushon, thoughts / interface?
1. should settings-backup also include ads ?
1. what buttons should show on this page, in what groups ?
1. should ad export/import/clear only happen in the vault ?

**Comment by cqx931** [comment]
How about showing the following buttons in settings page:
- Back up to file (add a dialogue box asking the user whether they want to include ads in the file)
- Restore from file
- Reset to default settings

And leave the ads only export/import/reset in the vault?

**Comment by dhowe** [comment]
I think that makes sense

**Comment by mushon** [comment]
Sorry for joining the party late, but @cqx931’s make sense to me. I think it would help clarify things.

Can we fin a way to address Halvor’s concerns too (made in a recent email):

> I log into various Oslo municipal library PCs several times a day. Every time I log in I have to install Ad Nauseam anew. Obviously, since these get wiped between logins.
>
> These all run Lubuntu 14.04.
>
> Once in a while I notice that after I have installed and enabled the add-on from the Firefox add-ons interface,
>
> the blackish  Ad Nauseam page with the various options
>
> does not open. Still Ad Nauseam seems to be working all the same, and the upper right hand menu looks like this:
>
> I just thought you might like to know about this observation.
>
> Halvor

I think what he needs is a way to install with custom settings. Or is there a simpler way to do it?

**Comment by dhowe** [comment]
Export a settings file before quitting, import it after a new install is the only way I can see...
@mushon, can you comment on the interface questions above?

**Comment by cqx931** [comment]
It makes sense to me to have three options in the dialog box, but the order would be:
1.only settings [default]
2.settings & ads
3.only ads

**Comment by dhowe** [comment]
And this would happen only in settings, while in the vault you could only export ads?

**Comment by cqx931** [comment]
yes, only exporting ads in the vault

**Comment by dhowe** [comment]
Yes, as soon as 3.5 is published, lets go ahead with what you suggest above

**Comment by cqx931** [comment]
Please check: https://github.com/dhowe/AdNauseam/pull/1312

**Comment by dhowe** [comment]
Yes, that makes sense to me

**Comment by dhowe** [comment]
See https://github.com/dhowe/AdNauseam/pull/1312#issuecomment-354266326

**Comment by dhowe** [comment]
@cqx931 where do we stand on this?

**Comment by cqx931** [comment]
I think this ticket can be closed as all has been implemented in #1312
